nitro-src/claim.ts

```diff
@@ -1,163 +1,219 @@
+ * @param targetOutcome initial outcome stored on chain for the target channel.
+ * @param exitRequest list of indices expressing which destinations in the allocation should be paid out for each asset.
+ */
 export function claim(
-  initialGuaranteeOutcome: Exit,
-  initialHoldings: BigNumberish[],
+  guarantees: Exit,
+  holdings: BigNumber[],
   targetChannelIndex: number,
-  initialTargetOutcome: Exit,
+  targetOutcome: Exit,
   exitRequest: number[][]
 ) {
-  if (initialTargetOutcome.length !== initialHoldings.length) throw Error;
-  const updatedTargetOutcome: Exit = [];
-  // We want to create a clone of the original values to prevent mutation
-  const updatedHoldings = initialHoldings.map((h) => BigNumber.from(h));
-  const updatedGuaranteeOutcome: Exit = JSON.parse(
-    JSON.stringify(initialGuaranteeOutcome)
-  );
-  const exit: Exit = [];
+  if (targetOutcome.length !== holdings.length) {
+    throw Error("targetOutcome length must match funds length");
+  }
 
-  // Iterate through every asset
-  for (
-    let assetIndex = 0;
-    assetIndex < initialGuaranteeOutcome.length;
-    assetIndex++
-  ) {
-    const guarantees = initialGuaranteeOutcome[assetIndex].allocations;
-    const targetAllocations = initialTargetOutcome[assetIndex].allocations;
+  // Deep-copy as we will mutating these objects.
+  const afterClaimGuarantee: Exit = _.cloneDeep(guarantees);
+  const afterClaimHoldings = _.cloneDeep(holdings);
+
+  const afterClaimTargetOutcome: Exit = [];
+  const afterClaimExits: Exit = [];
 
-    const updatedAllocations = targetAllocations.map((a) =>
-      JSON.parse(JSON.stringify(a))
-    ); // copy the allocations for mutation
+  // Iterate through every asset
+  for (let assetIndex = 0; assetIndex < guarantees.length; assetIndex++) {
+    const guaranteesForOneAsset = guarantees[assetIndex].allocations;
+    const targetAllocations = targetOutcome[assetIndex].allocations;
 
     const singleAssetExit: SingleAssetExit = {
-      ...initialTargetOutcome[assetIndex],
+      ...targetOutcome[assetIndex],
       allocations: [], // start with an empty array
     };
 
     if (
-      guarantees[targetChannelIndex].allocationType !== AllocationType.guarantee
-    )
-      throw Error;
+      guaranteesForOneAsset[targetChannelIndex].allocationType !==
+      AllocationType.guarantee
+    ) {
+      throw Error("Expected allocation type guarantee");
+    }
 
-    let surplus = BigNumber.from(initialHoldings[assetIndex]);
     // Any guarantees before this one have priority on the funds
     // So we must account for that by reducing the surplus
-    for (
-      let guaranteeIndex = 0;
-      guaranteeIndex < targetChannelIndex;
-      guaranteeIndex++
-    ) {
-      const { amount } = guarantees[guaranteeIndex];
+    const guaranteesBeforeOurs = guaranteesForOneAsset.slice(
+      0,
+      targetChannelIndex
+    );
+    const amountBeforeUs = sumAllocationAmounts(guaranteesBeforeOurs);
```

Comment:
```suggestion
    const allocationsBeforeOurs = guaranteesForOneAsset.slice(
      0,
      targetChannelIndex
    );
    const amountBeforeUs = sumAllocationAmounts(allocationsBeforeOurs);
```

Reply:
I think we probably want to introduce a unit test to make sure that we can handle allocations mixed with guarantees.